pvHelpers/misc.py

```diff
@@ -84,48 +100,103 @@ def updateLog(self, logobj):
+        calframe = inspect.getouterframes(curframe, 2)
+        try:
+            caller_info = self.__format_caller_info(calframe)
+            if self.logobj is not None:
+                self.logobj.info(" ".join([caller_info, str(string)]))
+            else:
+                print "{} INFO: {} {}".format(self.__now(), caller_info,
+                                              string)
+        finally:
+            del curframe
+            del calframe
 
     def warning(self, string):
-        if self.logobj is not None:
-            self.logobj.warning(string)
-        else:
-            print "{} WARN: {}".format(self.__now(), string)
-
-    warn=warning
+        curframe = inspect.currentframe()
+        calframe = inspect.getouterframes(curframe, 2)
+        try:
+            caller_info = self.__format_caller_info(calframe)
+            if self.logobj is not None:
+                self.logobj.warning(" ".join([caller_info, str(string)]))
+            else:
+                print "{} WARN: {} {}".format(self.__now(), caller_info,
+                                              string)
+        finally:
+            del curframe
+            del calframe
+
+    warn = warning
 
     def error(self, string):
-        if self.logobj is not None:
-            self.logobj.error(string)
-        else:
-            print "{} ERROR: {}".format(self.__now(), string)
+        curframe = inspect.currentframe()
+        calframe = inspect.getouterframes(curframe, 2)
+        try:
+            caller_info = self.__format_caller_info(calframe)
+            if self.logobj is not None:
+                self.logobj.error(" ".join([caller_info, str(string)]))
+            else:
+                print "{} ERROR: {} {}".format(self.__now(), caller_info,
+                                               string)
+        finally:
+            del curframe
+            del calframe
 
     def exception(self, exception):
-        if self.logobj is not None:
-            self.logobj.exception(exception)
-        else:
-            print "{} ERROR: {}".format(self.__now(), exception)
+        curframe = inspect.currentframe()
+        calframe = inspect.getouterframes(curframe, 2)
+        try:
+            caller_info = self.__format_caller_info(calframe)
+            if self.logobj is not None:
+                self.logobj.exception(" ".join([caller_info, str(exception)]))
+            else:
+                print "{} EXCEPTION: {} {}".format(self.__now(), caller_info,
+                                                   exception)
+        finally:
+            del curframe
+            del calframe
 
     def __now(self):
         return datetime.datetime.now().strftime("%Y/%m/%d %H:%M:%S")
 
+    def __format_caller_info(self, calframe):
+        """
+            :return: (filename, line_number, func)
+        """
+        filename = calframe[1][1]
+        line_number = calframe[1][2]
+        caller_func = calframe[1][3]
```

Comment:
what's the behavior if there's no caller_func ?

Reply:
added comment to clarify and also tests.